File: examples/running/campaign/campaign.go
package campaign

import "fmt"

const MaxShortLinksPerCampaign = 25

type Campaign struct {
	id    CampaignID
	name  CampaignName
	links []ShortLink
	_     [0]func()
}

type CampaignSpec struct {
	ID    string
	Name  string
	Links []ShortLinkSpec
}

func NewCampaign(spec CampaignSpec) (Campaign, error) {
	id, err := NewCampaignID(spec.ID)
	if err != nil {
		return Campaign{}, fmt.Errorf("invalid campaign id: %w", err)
	}
	name, err := NewCampaignName(spec.Name)
	if err != nil {
		return Campaign{}, fmt.Errorf("invalid campaign name: %w", err)
	}

	var links []ShortLink
	for i, linkSpec := range spec.Links {
		link, err := NewShortLink(linkSpec)
		if err != nil {
			return Campaign{}, fmt.Errorf("invalid short link at index %d: %w", i, err)
		}
		links, err = appendShortLink(links, link)
		if err != nil {
			return Campaign{}, fmt.Errorf("short link at index %d: %w", i, err)
		}
	}

	return Campaign{id: id, name: name, links: links}, nil
}

func (c Campaign) ID() CampaignID     { return c.id }
func (c Campaign) Name() CampaignName { return c.name }

func (c Campaign) Links() []ShortLink {
	out := make([]ShortLink, len(c.links))
	copy(out, c.links)
	return out
}

func (c *Campaign) AddShortLink(spec ShortLinkSpec) error {
	spec.Active = true
	link, err := NewShortLink(spec)
	if err != nil {
		return fmt.Errorf("invalid short link: %w", err)
	}
	links, err := appendShortLink(c.links, link)
	if err != nil {
		return err
	}
	c.links = links
	return nil
}

func (c *Campaign) DeactivateShortLink(slug Slug) error {
	for i := range c.links {
		if c.links[i].Slug() == slug {
			return c.links[i].Deactivate()
		}
	}
	return fmt.Errorf("no short link with slug %s in campaign %s", slug, c.id)
}

func appendShortLink(links []ShortLink, link ShortLink) ([]ShortLink, error) {
	if len(links) >= MaxShortLinksPerCampaign {
		return nil, fmt.Errorf("campaign already holds the maximum of %d short links", MaxShortLinksPerCampaign)
	}
	for _, existing := range links {
		if existing.Slug() == link.Slug() {
			return nil, fmt.Errorf("duplicate slug %s in campaign", link.Slug())
		}
	}
	return append(links, link), nil
}

File: examples/python/campaign/campaign.py
from dataclasses import dataclass

from campaign.campaign_id import CampaignID
from campaign.campaign_name import CampaignName
from campaign.short_link import ShortLink, ShortLinkSpec
from campaign.slug import Slug

MAX_SHORT_LINKS_PER_CAMPAIGN = 25


@dataclass(frozen=True)
class CampaignSpec:

    id: str
    name: str
    links: tuple[ShortLinkSpec, ...]


class Campaign:

    def __init__(self, spec: CampaignSpec) -> None:
        try:
            id = CampaignID(spec.id)
        except ValueError as e:
            raise ValueError(f"invalid campaign id: {e}") from e
        try:
            name = CampaignName(spec.name)
        except ValueError as e:
            raise ValueError(f"invalid campaign name: {e}") from e

        admitted: list[ShortLink] = []
        for i, link_spec in enumerate(spec.links):
            try:
                link = ShortLink(link_spec)
            except ValueError as e:
                raise ValueError(f"invalid short link at index {i}: {e}") from e
            admitted = _append_short_link(admitted, link)
        self._id = id
        self._name = name
        self._links = admitted

    @property
    def id(self) -> CampaignID:
        return self._id

    @property
    def name(self) -> CampaignName:
        return self._name

    @property
    def links(self) -> tuple[ShortLink, ...]:
        return tuple(link._clone() for link in self._links)

    def add_short_link(self, spec: ShortLinkSpec) -> None:
        spec = ShortLinkSpec(slug=spec.slug, target_url=spec.target_url, active=True)
        link = ShortLink(spec)
        self._links = _append_short_link(self._links, link)

    def deactivate_short_link(self, slug: Slug) -> None:
        for link in self._links:
            if link.slug == slug:
                link.deactivate()
                return
        raise ValueError(f"no short link with slug {slug} in campaign {self._id}")

    __eq__ = None  # type: ignore[assignment]
    __hash__ = None  # type: ignore[assignment]


def _append_short_link(links: list[ShortLink], link: ShortLink) -> list[ShortLink]:
    if len(links) >= MAX_SHORT_LINKS_PER_CAMPAIGN:
        raise ValueError(
            f"campaign already holds the maximum of "
            f"{MAX_SHORT_LINKS_PER_CAMPAIGN} short links"
        )
    for existing in links:
        if existing.slug == link.slug:
            raise ValueError(f"duplicate slug {link.slug} in campaign")
    return [*links, link]
